cpp/src/arrow/compute/exec/hash_join.cc

```diff
@@ -809,22 +1009,40 @@ class HashJoinBasicImpl : public HashJoinImpl {
                                       ScanHashTable_num_tasks());
   }
 
-  bool QueueBatchIfNeeded(int side, ExecBatch batch) {
+  Result<bool> QueueBatchIfNeeded(size_t thread_index, int side, ExecBatch& batch) {
     if (side == 0) {
-      std::lock_guard<std::mutex> lock(left_batches_mutex_);
-      if (has_hash_table_) {
-        return false;
+      // We don't want to do the filtering while holding the lock, since that can get
+      // expensive.
+      bool needs_filtering;
+      {
+        std::lock_guard<std::mutex> lock(left_batches_mutex_);
+        bloom_filters_ready_ = bloom_filters_ready_ || num_expected_bloom_filters_ == 0;
+        needs_filtering = bloom_filters_ready_ && num_expected_bloom_filters_ != 0;
       }
-      left_batches_.emplace_back(std::move(batch));
-      return true;
+      if (needs_filtering) RETURN_NOT_OK(ApplyBloomFiltersToBatch(thread_index, batch));
```

Comment:
```suggestion
      if (needs_filtering)
      {
          RETURN_NOT_OK(ApplyBloomFiltersToBatch(thread_index, batch));
          return false;
      }
```

Reply:
I don't think this is correct because we still want to enqueue the probe-side batch after we've filtered it.

Reply:
Ah, you're right.  I had thought the logic below was `queued = !bloom_filters_ready_ && !has_hash_table_;`